### PR TITLE
447 shuffle bug

### LINKED ISSUE #447: Bug Report: error when using `shuffle_ard()` after running `ard_stack_hierarchical()` with overall and over_variables = TRUE
### What happened?

I get an error when using `shuffle_ard()` after running `ard_stack_hierarchical()` with overall and over_variables = TRUE

### PR DESCRIPTION
**Pull request by bzkrouse**
**What changes are proposed in this pull request?**
* Style this entry in a way that can be copied directly into `NEWS.md`. (#<issue number>, @<username>)

Provide more detail here as needed.

**Reference GitHub issue associated with pull request.** _e.g., 'closes #<issue number>'_
closes #447 

--------------------------------------------------------------------------------

Pre-review Checklist (if item does not apply, mark is as complete)
- [ ] **All** GitHub Action workflows pass with a :white_check_mark:
- [ ] PR branch has pulled the most recent updates from master branch: `usethis::pr_merge_main()`
- [ ] If a bug was fixed, a unit test was added.
- [ ] Code coverage is suitable for any new functions/features (generally, 100% coverage for new code): `devtools::test_coverage()`
- [ ] Request a reviewer

Reviewer Checklist (if item does not apply, mark is as complete)

- [ ] If a bug was fixed, a unit test was added.
- [ ] Run `pkgdown::build_site()`. Check the R console for errors, and review the rendered website.
- [ ] Code coverage is suitable for any new functions/features: `devtools::test_coverage()`

When the branch is ready to be merged:
- [ ] Update `NEWS.md` with the changes from this pull request under the heading "`# cards (development version)`". If there is an issue associated with the pull request, reference it in parentheses at the end update (see `NEWS.md` for examples).
- [ ] **All** GitHub Action workflows pass with a :white_check_mark:
- [ ] Approve Pull Request
- [ ] Merge the PR. Please use "Squash and merge" or "Rebase and merge".

_Optional Reverse Dependency Checks_:

Install `checked` with `pak::pak("Genentech/checked")` or `pak::pak("checked")`

```shell
# Check dev versions of `cardx`, `gtsummary`, and `tfrmt` which are in the `ddsjoberg` R Universe
Rscript -e "options(checked.check_envvars = c(NOT_CRAN = TRUE)); checked::check_rev_deps(path = '.', n = parallel::detectCores() - 2L, repos = c('https://ddsjoberg.r-universe.dev', 'https://cloud.r-project.org'))"

# Check CRAN reverse dependencies but run tests skipped on CRAN
Rscript -e "options(checked.check_envvars = c(NOT_CRAN = TRUE)); checked::check_rev_deps(path = '.', n = parallel::detectCores() - 2, repos = 'https://cloud.r-project.org')"

# Check CRAN reverse dependencies in a CRAN-like environment
Rscript -e "options(checked.check_envvars = c(NOT_CRAN = FALSE), checked.check_build_args = '--as-cran'); checked::check_rev_deps(path = '.', n = parallel::detectCores() - 2, repos = 'https://cloud.r-project.org')"
```

### CODE REVIEW

**Review by dragosmg — COMMENTED** [review]
@bzkrouse Looks good. I do not have many comments to add. You mentioned that a more elaborated logic in the mutate across function is needed due to an edge case (i.e. when `Overall <col>` already exists. Should we add a test for that. Could be useful (both now and if we refactor at a later date).

File: R/shuffle_ard.R [hunk 1]
```diff
@@ -44,19 +46,19 @@ shuffle_ard <- function(x, trim = TRUE) {
...
+    dplyr::select(-all_of(vars_protected), "..cards_idx..")
 
   dat_cards_stats <- dat_cards |>
     dplyr::select(all_of(vars_protected))
 
   # process the data/variable info
   dat_cards_grps_processed <- dat_cards_grps |>
     .check_var_nms(vars_protected = names(dat_cards_stats)) |>
-    rename_ard_columns(columns = all_ard_groups("names"), fill = "Overall {colname}") |>
+    rename_ard_columns(columns = all_ard_groups("names"), fill = "..cards_overall..") |>
```
Comment (dragosmg) [inline R/shuffle_ard.R:57]: I might be misunderstanding the intention behind `rename_ard_columns()`, more specifically its `fill` argument. AFAIU , in this context, the call should look at any columns named `groupx`,  etc. and their corresponding `groupx_level`, etc. If `groupx` is `NA` and `groupx_level` is `NULL` then the new column (called `<var>`) should get the fill value (in our case `"..cards_overall.."`). However, it does not seem to achieve that. 

I guess the crux is in the docs. What does:

> fill (scalar/glue) a scalar to fill column values when the variable does not have levels.

mean? Is the `fill` value used only when the variable does not have _any_ levels or when the specific entry does not have a matching / corresponding level?

If it's the former, the behaviour is correct, but understanding its impact is tricky. And `fill` will effectively only apply to non-categorical variables (most often continuous). 

If it's the latter (where I think my understanding lies) then it might be a bug.

Comment (bzkrouse) [reply R/shuffle_ard.R:57]: @dragosmg  The fill value is used anytime the variable's level is NULL, which is covered by this unit test - https://github.com/insightsengineering/cards/blob/main/tests/testthat/test-shuffle_ard.R#L157 

Do you think another is needed?

**Comment by bzkrouse** [comment]
> @bzkrouse Looks good. I do not have many comments to add. You mentioned that a more elaborated logic in the mutate across function is needed due to an edge case (i.e. when `Overall <col>` already exists. Should we add a test for that? Could be useful (both now and if we refactor at a later date).

thank you @dragosmg for the review! Added a test (and fixed a bit of broken logic too). Feel free to have another look.

File: R/shuffle_ard.R (hunk 1 above)
Comment (dragosmg) [reply R/shuffle_ard.R:57]: @bzkrouse  It looks to me like that unit test has a slightly malformed input (`group1` is made up of `ARM` and `SEX`) :
```r
> ard_cardx
{cards} data frame: 4 x 9
  group1 variable   context stat_name stat_label  stat
1    ARM   AGEGR1 stats_ch… statistic  X-square… 5.079
2    ARM   AGEGR1 stats_ch…   p.value    p-value 0.079
3    SEX   AGEGR1 stats_ch… statistic  X-square… 1.039
4    SEX   AGEGR1 stats_ch…   p.value    p-value 0.595
ℹ 3 more variables: fmt_fun, warning, error
```

Probably a more realistic `card` would look like:
```r
> ard_cardx
{cards} data frame: 4 x 10
  group1 group2 variable stat_name stat_label  stat
1    ARM   <NA>   AGEGR1 statistic  X-square… 5.079
2    ARM   <NA>   AGEGR1   p.value    p-value 0.079
3   <NA>    SEX   AGEGR1 statistic  X-square… 1.039
4   <NA>    SEX   AGEGR1   p.value    p-value 0.595
ℹ 4 more variables: context, fmt_fun, warning, error
```

but even the second one looks somewhat unrealistic (to my untrained eye). I struggle a bit to re-construct the call that lead to the object. How can anyone group by one or more categorical variables with no levels? Shouldn't there be at least 1 level in either of the grouping variables?

The first `card` looks like it was created by grouping on a variable with 2 levels (`"ARM"` and `"SEX`), but again that would produce a `group1_level` column, wouldn't it?

Anyways, we are probably in the first scenario (out of the 2 I mentioned above): the `fill` value is used **only** when the variable does not have any levels (e.g. `group1` exists and the `group1_level` column is either completely empty - `NULL` - or it does not exist). 

I have worked through multiple `shuffle_ard()` examples, and I still cannot find one in which this line has any impact:
https://github.com/insightsengineering/cards/blob/149c41850aca169113e4594faddc502a7687d070/R/shuffle_ard.R#L55

Happy to approve as my comments are somewhat inconsequential to the PR.

**Review by dragosmg — APPROVED** [review]
Many thanks. I left a couple of comments. They do not need to be addressed, they're more conversation points.

File: R/shuffle_ard.R (hunk 1 above)
Comment (bzkrouse) [reply R/shuffle_ard.R:57]: @dragosmg this will occur if you combine cardx outputs, such as chisq tests with different `by` variable ( this is what that unit test reference above is simulating). Happy to chat more about this!

**Review by ddsjoberg — APPROVED** [review]
Thanks!!
